src/pprof/ddprof_pprof.cc

```diff
@@ -91,38 +91,37 @@ DDRes pprof_create_profile(DDProfPProf *pprof, DDProfContext *ctx) {
     }
   }
 
-  // If none of the samples were good, that's an error
-  if (!num_sample_type_ids) {
-    // We use the phrase "profile type" in the error, since this is more
-    // obvious for customers.
-    DDRES_RETURN_ERROR_LOG(DD_WHAT_PPROF, "No valid profile types given");
-  }
-
   pprof->_nb_values = num_sample_type_ids;
   ddprof_ffi_Slice_value_type sample_types = {.ptr = perf_value_type,
                                               .len = pprof->_nb_values};
 
-  // Populate the default.  If we have a frequency, assume it is given in hertz
-  // and convert to a period in nanoseconds.  This is broken for many event-
-  // based types (but providing frequency would also be broken in those cases)
-  int64_t default_period = default_watcher->sample_period;
-  if (default_watcher->options.is_freq)
-    default_period = 1e9 / default_period;
-
-  ddprof_ffi_Period period = {
-      .type_ = perf_value_type[pv[default_watcher->pprof_sample_idx]],
-      .value = default_period,
-  };
-
-  pprof->_profile = ddprof_ffi_Profile_new(sample_types, &period);
+  ddprof_ffi_Period period;
+  if (num_sample_type_ids > 0) {
+    // Populate the default.  If we have a frequency, assume it is given in
+    // hertz and convert to a period in nanoseconds.  This is broken for many
+    // event- based types (but providing frequency would also be broken in those
+    // cases)
+    int64_t default_period = default_watcher->sample_period;
+    if (default_watcher->options.is_freq)
+      default_period = 1e9 / default_period;
+
+    period = {
+        .type_ = perf_value_type[pv[default_watcher->pprof_sample_idx]],
+        .value = default_period,
+    };
+  }
+  pprof->_profile = ddprof_ffi_Profile_new(
+      sample_types, num_sample_type_ids > 0 ? &period : nullptr);
   if (!pprof->_profile) {
-    DDRES_RETURN_ERROR_LOG(DD_WHAT_PPROF, "Unable to allocate profiles");
+    DDRES_RETURN_ERROR_LOG(DD_WHAT_PPROF, "Unable to create profile");
   }
   return ddres_init();
 }
 
 DDRes pprof_free_profile(DDProfPProf *pprof) {
-  ddprof_ffi_Profile_free(pprof->_profile);
+  if (pprof->_profile) {
```

Comment:
Nit: does this check actually improve readability?  `free(NULL)` is valid as per POSIX (https://pubs.opengroup.org/onlinepubs/009604499/functions/free.html)

Reply:
Well Rust has a different opinion on the matter, `ddprof_ffi_Profile_free` crashes if passed a null pointer.
I had a crash when `pprof_create_profile` was failing and `pprof_free_profile` was subsequently trying to free a null profile (just try using `-e sDUM` alone on main).

Reply:
Interesting !